src/mscorlib/src/System/Lazy.cs

```diff
@@ -27,12 +27,158 @@
+        /// </summary>
+        /// <param name="state"></param>
+        internal LazyNonGeneric(LazyState state)
+        {
+            State = state;
+        }
+
+        private Exception InvalidLogic()
+        {
+            // correctly implemented, we should never create this exception
+            return new Exception("Invalid logic; execution should not get here");
+        }
+
+        /// <summary>
+        /// Constructor used for exceptions
+        /// </summary>
+        /// <param name="mode"></param>
+        /// <param name="exception"></param>
+        internal LazyNonGeneric(LazyThreadSafetyMode mode, Exception exception)
+        {
+            if (mode == LazyThreadSafetyMode.ExecutionAndPublication)
+                State = LazyState.ExecutionAndPublicationException;
+            else if (mode == LazyThreadSafetyMode.None)
+                State = LazyState.NoneException;
+            else if (mode == LazyThreadSafetyMode.PublicationOnly)
+                State = LazyState.PublicationOnlyException;
+            else
+                throw InvalidLogic();
+
+            _exceptionDispatch = System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(exception);
+        }
+
+        internal void ThrowException()
+        {
+            if (_exceptionDispatch == null)
+                throw InvalidLogic();
+
+            _exceptionDispatch.Throw();
+        }
+
+        private LazyThreadSafetyMode GetMode()
+        {
+            switch (State)
+            {
+                case LazyState.NoneViaConstructor:
+                case LazyState.NoneViaFactory:
+                case LazyState.NoneException:
+                    return LazyThreadSafetyMode.None;
+
+                case LazyState.PublicationOnlyViaConstructor:
+                case LazyState.PublicationOnlyViaFactory:
+                case LazyState.PublicationOnlyWait:
+                case LazyState.PublicationOnlyException:
+                    return LazyThreadSafetyMode.PublicationOnly;
+
+                case LazyState.ExecutionAndPublicationViaConstructor:
+                case LazyState.ExecutionAndPublicationViaFactory:
+                case LazyState.ExecutionAndPublicationException:
+                    return LazyThreadSafetyMode.ExecutionAndPublication;
+
+                default:
+                    throw InvalidLogic();
+            }
+        }
+
+        internal static LazyThreadSafetyMode GetMode(LazyNonGeneric state)
+        {
+            if (state == null)
+                return LazyThreadSafetyMode.None; // we don't know the mode anymore
+            return state.GetMode();
+        }
+
+        private bool GetIsValueFaulted()
+        {
+            return _exceptionDispatch != null;
+        }
+
+        internal static bool GetIsValueFaulted(LazyNonGeneric state)
+        {
+            if (state == null)
+                return false;
+            return state.GetIsValueFaulted();
+        }
```

Comment:
There's no need for the first method above to be separated out; the second member can just be:
```C#
internal static bool GetIsValueFaulted() => state?._exceptionDispatch != null;
```